pkg/kubelet/cm/topologymanager/policy_single_numa_node.go

```diff
@@ -53,12 +55,134 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
+				!hints[res].NUMANodeAffinity.IsEqual(candidate.NUMANodeAffinity) {
+				// hint not found, move to next hint from allResourcesHints[0]
+				foundMatch = false
+				break
+			}
+		}
+		if foundMatch {
+			match = candidate
+			match.Preferred = true
+			break
+		}
+	}
+	return foundMatch, match
+}
+
+// Return hints that have valid bitmasks with exactly one bit set
+func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
+	var filteredResourcesHints [][]TopologyHint
+	if len(allResourcesHints) > 0 {
+		for _, oneResourceHints := range allResourcesHints {
+			var filtered []TopologyHint
+			if len(oneResourceHints) > 0 {
+				for _, hint := range oneResourceHints {
+					if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 {
+						filtered = append(filtered, hint)
+					}
+				}
+			}
+			filteredResourcesHints = append(filteredResourcesHints, filtered)
+		}
+	}
+	return filteredResourcesHints
+}
+
 func (p *singleNumaNodePolicy) mergeProvidersHints(providersHints []map[string][]TopologyHint) TopologyHint {
 	// Set the default hint to return from this function as an any-NUMANode
 	// affinity with an unpreferred allocation. This will only be returned if
 	// no better hint can be found when merging hints from each hint provider.
 	defaultAffinity, _ := bitmask.NewBitMask(p.numaNodes...)
 	defaultHint := TopologyHint{defaultAffinity, false}
+
+	// Loop through all provider hints and save an accumulated list of the
+	// hints returned by each hint provider. If no hints are provided, assume
+	// that provider has no preference for topology-aware allocation.
+	var allResourcesHints [][]TopologyHint
+	for _, hints := range providersHints {
+		if len(hints) == 0 {
+			klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with any resource, " +
+				"skipping.")
+			continue
+		}
+
+		// Otherwise, accumulate the hints for each resource type into allProviderHints.
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for NUMA affinity with resource "+
+					"'%s', skipping.", resource)
+				continue
+			}
+
+			if len(hints[resource]) == 0 {
+				klog.Infof("[topologymanager] Hint Provider has no possible NUMA affinities for resource '%s'",
+					resource)
+				// return defaultHint which will fail pod admission
+				return defaultHint
+			}
+			klog.Infof("[topologymanager] TopologyHints for resource '%v': %v", resource, hints[resource])
+			allResourcesHints = append(allResourcesHints, hints[resource])
+		}
+	}
+	// In case allProviderHints length is zero it means that we have no
+	// preference for NUMA affinity. In that case update default hint preferred
+	// to true to allow scheduling.
+	if len(allResourcesHints) == 0 {
+		klog.Infof("[topologymanager] No preference for NUMA affinity from all providers")
+		defaultHint.Preferred = true
+		return defaultHint
+	}
```

Comment:
It might be good to expand this comment a bit since we are in the `single-numa-node` policy, and it seems counter-intuitive on the surface to ever allow scheduling if we can't align on a single NUMA node. We should make it clear, however, that we need someway of scheduling pods that don't require *any* aligned resources, which is why we have this special-case. The policy itself only ever kicks in, once some resource requires alignment.

That said, I would expect to return `TopologyHint{nil, true}` here, and not the `defaultHint` as defined here.